android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreAesCbc.java

```diff
@@ -0,0 +1,236 @@
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.spec.KeySpec;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.crypto.KeyGenerator;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+
+/**
+ * @see <a href="https://proandroiddev.com/secure-data-in-android-initialization-vector-6ca1c659762c">Secure Data in Android</a>
+ * @see <a href="https://stackoverflow.com/questions/36827352/android-aes-with-keystore-produces-different-cipher-text-with-same-plain-text">AES cipher</a>
+ */
+@TargetApi(Build.VERSION_CODES.M)
+@SuppressWarnings({"unused", "WeakerAccess"})
+public class CipherStorageKeystoreAesCbc extends CipherStorageBase {
+  //region Constants
+  /** AES */
+  public static final String ALGORITHM_AES = KeyProperties.KEY_ALGORITHM_AES;
+  /** CBC */
+  public static final String BLOCK_MODE_CBC = KeyProperties.BLOCK_MODE_CBC;
+  /** PKCS7 */
+  public static final String PADDING_PKCS7 = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+  /** Transformation path. */
+  public static final String ENCRYPTION_TRANSFORMATION =
+    ALGORITHM_AES + "/" + BLOCK_MODE_CBC + "/" + PADDING_PKCS7;
+  /** Key size. */
+  public static final int ENCRYPTION_KEY_SIZE = 256;
+
+  public static final String DEFAULT_SERVICE = "RN_KEYCHAIN_DEFAULT_ALIAS";
+  //endregion
+
+  //region Configuration
+  @Override
+  public String getCipherStorageName() {
+    return KnownCiphers.AES;
+  }
+
+  /** API23 is a requirement. */
+  @Override
+  public int getMinSupportedApiLevel() {
+    return Build.VERSION_CODES.M;
+  }
+
+  /** it can guarantee security levels up to SECURE_HARDWARE/SE/StrongBox */
+  @Override
+  public SecurityLevel securityLevel() {
+    return SecurityLevel.SECURE_HARDWARE;
+  }
+
+  /** Biometry is Not Supported. */
+  @Override
+  public boolean isBiometrySupported() {
+    return false;
+  }
+
+  /** AES. */
+  @Override
+  @NonNull
+  protected String getEncryptionAlgorithm() {
+    return ALGORITHM_AES;
+  }
+
+  /** AES/CBC/PKCS7Padding */
+  @NonNull
+  @Override
+  protected String getEncryptionTransformation() {
+    return ENCRYPTION_TRANSFORMATION;
+  }
+
+  /** {@inheritDoc}. Override for saving the compatibility with previous version of lib. */
+  @Override
+  public String getDefaultAliasServiceName() {
+    return DEFAULT_SERVICE;
+  }
+
+  //endregion
+
+  //region Overrides
+  @Override
+  @NonNull
+  public EncryptionResult encrypt(@NonNull final String alias,
+                                  @NonNull final String username,
+                                  @NonNull final String password,
+                                  @NonNull final SecurityLevel level)
+    throws CryptoFailedException {
+
+    throwIfInsufficientLevel(level);
+
+    final String safeAlias = getDefaultAliasIfEmpty(alias, getDefaultAliasServiceName());
+    final AtomicInteger retries = new AtomicInteger(1);
```

Comment:
just wondering, why is `AtomicInteger` needed?

Reply:
we send counter as a reference to another method that actually doing the loop based on this counter... needed boxing of the value. I have a choice: AtomicInteger vs Integer ... atomic types give a little better API. (from a performance perspective it's not a big loss)